### ISSUE TITLE
Dilated convolution pass not working on standard TCN model

### ISSUE THREAD

**Issue by bigcola317** [tensorflow/tensorflow#38638]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- TensorFlow installed from (source or binary): source
- TensorFlow version (or github SHA if from source): 17dfa4e121c080a547e9cf6443b8fe2ae9ed45ed

**Command used to run the converter or code if you’re using the Python API**
If possible, please share a link to Colab/Jupyter/any notebook.

```
# Copy and paste here the exact command (in Python)
model = load_model(model_path, custom_objects={'TCN': TCN})
converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.experimental_new_converter = True
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]
```
```
# Copy and paste here the exact command (tf_tfl_translate command-line utility)
tf_tfl_translate  --tf-input-arrays=input_1 --tf-input-shapes=1,784,1 --tf-output-arrays=dense/BiasAdd --print-function-result-mapping -o=$HOME/Desktop/converted.tflite --emit-builtin-tflite-ops ~/smnist.pb
```
**The output from the converter invocation**

```
# Copy and paste the output here. (tf_tfl_translate output)
'main' inputs:
	name: 'input_1' buffer: 0
'main' outputs:
	name: 'dense/BiasAdd' buffer: 271 loc("dense/BiasAdd")
```

**Also, please include a link to the saved model or GraphDef**
[smnist.zip](https://github.com/tensorflow/tensorflow/files/4492741/smnist.zip) contains the model in both .h5 and .pb format, and the converted Flatbuffers.

**Failure details**
I'm converting the Keras-TCN model from https://github.com/philipperemy/keras-tcn to TF Lite.

Method 1:
I convert the Keras model (.h5) to a TF Lite Flatbuffer via the Python API as above. The conversion is successful, inference with the model has the correct accuracy, but I want to get rid of the SpaceToBatchNd and BatchToSpaceNd nodes resulting from the Conv1D ops present in the model. Commit f54bb6f5578b931d79884302768996ba1073f685 claims to do so solving issue #29509. However this does not happen to be so in my converted model, which I attach.

Method 2:
To investigate further, I built the tf_tfl_translate tool from source and invoked it with the command above on a GraphDef (.pb) of the model. Conversions happen to be successful but again the STB and BTS ops are still present in the TF Lite Flatbuffer. The sequence of ops in the GraphDef seems to comply to the sequences specified in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/transforms/dilated_conv.h. 
However, this time the ExpandDims and Squeeze ops are correctly converted to Reshape operations, as specified here https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td, whereas this would not happen with the method above, and although I would like to avoid this being done on my final Flatbuffer.

**Comment by haozha111** [comment]
Hi,

The current dilated conv rewriting pass only supports Conv2D:
https://github.com/tensorflow/tensorflow/blob/b422faa5e39f0ea2e9502ed5d974f23c36768894/tensorflow/compiler/mlir/lite/transforms/prepare_tf.cc#L629

If your model contains Conv1D, unfortunately that's not supported now.

**Comment by bigcola317** [comment]
Thank you for your interest,

But actually the model presents exactly a Conv2D op, as the Conv1D is converted to an ExpandDims -> Conv2D -> Squeeze sequence. So in the frozen graph there is exactly a SpaceToBatchND -> Expand -> Conv2D -> Squeeze -> BatchToSpaceND -> BiasAdd sequence which this comment states is supported
https://github.com/tensorflow/tensorflow/blob/adbacb4206cf440f3628e7240a53d8e0f12122b6/tensorflow/compiler/mlir/lite/transforms/dilated_conv.h#L50
The same comment cites Wavenet as a use case for this sequence, which is very similar to my model.

**Comment by haozha111** [comment]
I took a look at your graph. And I think the pattern doesn't get matched because this line:
https://github.com/tensorflow/tensorflow/blob/adbacb4206cf440f3628e7240a53d8e0f12122b6/tensorflow/compiler/mlir/lite/transforms/dilated_conv.h#L284

For example, tensor name: tcn/residual_block_1_1/conv1D_1/SpaceToBatchND/block_shape only has 1 element '2', so the pattern got rejected since it's less than 2 elements.

Could you change your model code where you specify the `block_shape` argument for the SpaceToBatchND op? Assuming you want to apply dilation rate 2 on both the height and width, you can just pass in [2, 2] in this case.

**Comment by bigcola317** [comment]
Thank you.

The problem is that I didn't create the ExpandDims -> Conv2D -> Squeeze sequence explicitly, so I don't have direct access to them. I have dilated Conv1D layers in my model which are handled by TensorFlow this way. The complexity that TF introduces to handle Conv1D is actually quite surprising. So I think that the conversion should take care of supporting dilated Conv1D as it does dilated Conv2D. I could extend it myself in a pull request if that could help.

**Comment by haozha111** [comment]
I see. I think the TF high level python API just rewrite the dilated conv into SpaceToBatch -> Conv -> BatchToSpace automatically. If you could help extend this pass by supporting Conv1D that will be great. Thanks!

**Comment by bigcola317** [comment]
Hi @haozha111,

Unfortunately I cannot make it to work on this, is there a chance someone could solve this?
I kind of depend on it, but if this weren't possible I'd go with an ugly custom pass in my hardware delegates for turnaround time reasons.

I would be available to support with more information on how to reproduce the issue, etc.

**Comment by haozha111** [comment]
Hi,

I wouldn't have much time working on support for dilated conv1d. Do you mind either file a github issue or write the custom pass on your own? I believe the work should be  very similar with the current 2d case with little modifications.
